omnigibson/transition_rules.py

```diff
@@ -20,37 +40,205 @@
+        return dict(individual=individual_candidates, group=group_candidates)
 
     @classmethod
-    def update(cls):
+    def prune_active_rules(cls, objects):
         """
-        Updates the internal state by checking the filter status on all filter inputs
+        Prunes the active transition rules, removing any whose filter requirements are not satisifed by object set
+        @objects. Useful when the current object set changes, e.g.: an object is removed from the simulator
+
+        Args:
+            objects (list of BaseObject): List of objects that will be used to infer which transition rules are
+                currently active
         """
-        raise NotImplementedError()
+        # Need explicit list to iterate over because refresh_rules mutates the ACTIVE_RULES dict in place
+        cls.refresh_rules(rules=list(cls.ACTIVE_RULES.keys()), objects=objects)
 
+    @classmethod
+    def refresh_all_rules(cls, objects):
+        """
+        Refreshes all registered rules given the current set of objects @objects
+
+        Args:
+            objects (list of BaseObject): List of objects that will be used to infer which transition rules are
+                currently active
+        """
+        global RULES_REGISTRY
+
+        # Clear all active rules
+        cls.ACTIVE_RULES = dict()
+
+        # Refresh all registered rules
+        cls.refresh_rules(rules=RULES_REGISTRY.objects, objects=objects)
+
+    @classmethod
+    def refresh_rules(cls, rules, objects):
+        """
+        Refreshes the specified transition rules @rules based on current set of objects @objects. This will prune
+        any pre-existing rules in cls.ACTIVE_RULES if no valid candidates are found, or add / update the entry if
+        valid candidates are found
+
+        Args:
+            rules (list of BaseTransitionRule): List of transition rules whose candidate lists should be refreshed
+            objects (list of BaseObject): List of objects that will be used to infer which transition rules are
+                currently active
+        """
+        for rule in rules:
+            # Check if rule is still valid, if so, update its entry
+            candidates = cls.get_rule_candidates(rule=rule, objects=objects)
+
+            # Update candidates if valid, otherwise pop the entry if it exists in cls.ACTIVE_RULES
+            if candidates is not None:
+                # We have a valid rule which should be active, so grab and initialize all of its conditions
+                rule.initialize()
+                conditions = []
+                combined_candidates = dict(**candidates["individual"], **candidates["group"])
+                for condition in rule.conditions:
+                    condition.initialize(combined_candidates)
+                    conditions.append(condition)
+                cls.ACTIVE_RULES[rule] = {
+                    "conditions": conditions,
+                    "candidates": candidates,
+                }
```

Comment:
hmm is this logic correct? What is `rule` is already in the `ACTIVE_RULES`? 


```
for rule in rules:
    # Check if rule is still valid, if so, update its entry
    candidates = cls.get_rule_candidates(rule=rule, objects=objects)

    if candidates is None and rule in cls.ACTIVE_RULES:
        cls.ACTIVE_RULES.pop(rule)
    elif candidates is not None and rule not in cls.ACTIVE_RULES:
        rule.initialize()
        conditions = []
        combined_candidates = dict(**candidates["individual"], **candidates["group"])
        for condition in rule.conditions:
            condition.initialize(combined_candidates)
            conditions.append(condition)
        cls.ACTIVE_RULES[rule] = {
            "conditions": conditions,
            "candidates": candidates,
        }
```

Also, I am worried that a rule might be initialized more than once?